plugins/docker/attacher.go

```diff
@@ -81,52 +70,142 @@ func NewAttachManager(endpoint, certPath string, attachErrors chan<- error, name
 	if err != nil {
 		return nil, err
 	}
 
 	m := &AttachManager{
-		attached:      make(map[string]*LogPump),
-		channels:      make(map[chan *AttachEvent]struct{}),
-		client:        client,
-		errors:        attachErrors,
-		events:        make(chan *docker.APIEvents),
-		nameFromEnv:   nameFromEnv,
-		fieldsFromEnv: fieldsFromEnv,
+		client:           client,
+		events:           make(chan *docker.APIEvents),
+		nameFromEnv:      nameFromEnv,
+		fieldsFromEnv:    fieldsFromEnv,
+		fieldsFromLabels: fieldsFromLabels,
 	}
 
-	// Attach to all currently running containers
-	if containers, err := client.ListContainers(docker.ListContainersOptions{}); err == nil {
-		for _, listing := range containers {
-			m.attach(listing.ID[:12])
+	return m, nil
+}
+
+// Handler to wrap functions with retry logic
+func withRetries(doWork func() error) error {
+	var err error
+	for i := 0; i < CONNECT_RETRIES; i++ {
+		err := doWork()
+		if err == nil {
+			return nil
 		}
-	} else {
-		return nil, err
+		time.Sleep(retryInterval[i] * time.Second)
 	}
 
-	if err := client.AddEventListener(m.events); err != nil {
-		return nil, err
+	return err
+}
+
+// Attach to all running containers
+func (m *AttachManager) attachAll() error {
+	containers, err := m.client.ListContainers(docker.ListContainersOptions{})
+	if err != nil {
+		m.ir.LogError(err)
+		return err
 	}
 
-	go m.recvDockerEvents()
-	return m, nil
+	for _, listing := range containers {
+		err := withRetries(func() error {
+			return m.attach(listing.ID[:12], m.client)
+		})
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Main body of work
+func (m *AttachManager) Run(ir InputRunner, hostname string, stopChan chan error) error {
+	m.ir = ir
+	m.hostname = hostname
+
+	// Retry this up to CONNECT_RETRIES number of times, sleeping
+	// the defined interval. During this time the Docker client should
+	// get reconnected if there is a general connection issue.
+	err := withRetries(m.attachAll)
+	if err != nil {
+		m.ir.LogError(err)
+		m.ir.LogError(
+			fmt.Errorf("Failed to attach to Docker containers after %s retries. Plugin giving up.", CONNECT_RETRIES),
+		)
+		return err
+	}
+
+	m.ir.LogMessage("Attached to all containers")
+
+	err = withRetries(func() error { return m.client.AddEventListener(m.events) })
+	if err != nil {
+		m.ir.LogError(
+			fmt.Errorf("Failed to add Docker event listener after %s retries. Plugin giving up.", CONNECT_RETRIES),
+		)
+		return err
+	}
+
+	m.handleDockerEvents(stopChan)
+	return nil
+}
+
+// Try to reboot all of our connections
+func (m *AttachManager) restart() {
+	var err error
+
+	m.ir.LogMessage("Lost connection to Docker, re-connecting")
+	m.client.RemoveEventListener(m.events)
+	m.events = make(chan *docker.APIEvents) // RemoveEventListener closes it
+
+	m.client.AddEventListener(m.events)
+
+	err = withRetries(m.attachAll)
+	if err != nil {
+		m.ir.LogError(fmt.Errorf("Unable to attach to some containers! (%s)", err.Error()))
```

Comment:
I'm a bit confused. If `withRetries(m.attachAll)` returns an error in the Run method, then Run itself returns the error which AFAICT will cause the plugin to exit. But if it returns an error here in the restart method, then we just log the error and take no further action, everything keeps running as usual. Am I understanding this correctly? Is this actually the behaviour that we want?